### PR TITLE
Add support for HubertForCTC

### PR DESCRIPTION
**Pull request by jimypbr**
# What does this PR do?

Adds support for `HubertForCTC` model.

Can fine-tune it on librispeech with the command:
```
python examples/speech-recognition/run_speech_recognition_ctc.py \
    --dataset_name "librispeech_asr" \
    --dataset_config_name "clean" \
    --train_split_name "train.100" \
    --eval_split_name "validation" \
    --model_name_or_path facebook/hubert-base-ls960 \
    --ipu_config_name Graphcore/wav2vec2-ctc-base-ipu \
    --ipu_config_overrides "device_iterations=16,inference_device_iterations=16" \
    --mask_time_prob 0.0 \
    --mask_feature_prob 0.0 \
    --output_dir "./hubert-base-960h" \
    --overwrite_output_dir \
    --length_column_name "input_length" \
    --num_train_epochs 5 \
    --learning_rate "3e-4" \
    --warmup_steps 400 \
    --evaluation_strategy "steps" \
    --text_column_name "text" \
    --save_steps 400 \
    --eval_steps 400 \
    --logging_steps 10 \
    --save_total_limit 1 \
    --freeze_feature_encoder \
    --do_train \
    --do_eval \
    --layerdrop 0.0 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 2 \
    --pod_type pod8 \
    --adam_beta1 0.9 \
    --adam_beta2 0.98 \
    --adam_epsilon 0.0001 \
    --report_to wandb \
    --dataloader_drop_last \
    --dataloader_mode "async_rebatched" \
    --dataloader_num_workers 8
```

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/models/hubert/modeling_hubert.py [hunk 1]
```diff
@@ -67,3 +73,180 @@ def deparallelize(self):
...
+    def _add_begin_block(self, module, name, ipu_id):
+        poptorch.BeginBlock(module, name, ipu_id)
+
+    def parallelize(self):
+        super().parallelize()
+
+        self.freeze_feature_encoder()
+        self.change_hubert_encoder_class(False)
+
+        # self.hubert.feature_extractor = poptorch.BeginBlock(self.hubert.feature_extractor, ipu_id=0)
```
Comment (katalinic-gc) [inline optimum/graphcore/models/hubert/modeling_hubert.py:101]: should this commented out section be removed?

File: optimum/graphcore/models/hubert/modeling_hubert.py [hunk 2]
```diff
@@ -67,3 +73,180 @@ def deparallelize(self):
...
+            # Project Hidden
+            layers.append(("Project Hidden", self.lm_head))
+
+            layer_ipu = get_layer_ipu(self.ipu_config, layers)
+
+            for i, (name, layer) in enumerate(layers):
+                logger.info(f"{name} --> IPU {layer_ipu[i]}")
+                self._add_begin_block(layer, name, ipu_id=layer_ipu[i])
+
+            logger.info("---------------------------------------")
```
Comment (katalinic-gc) [inline optimum/graphcore/models/hubert/modeling_hubert.py:137]: return self at the end

File: optimum/graphcore/models/hubert/modeling_hubert.py [hunk 3]
```diff
@@ -67,3 +73,180 @@ def deparallelize(self):
...
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        hidden_states = outputs[0]
+        hidden_states = self.dropout(hidden_states)
+
+        logits = self.lm_head(hidden_states)
+
+        # loss = None
```
Comment (katalinic-gc) [inline optimum/graphcore/models/hubert/modeling_hubert.py:179]: likewise re removing commented out blocks

File: optimum/graphcore/models/hubert/modeling_hubert.py [hunk 4]
```diff
@@ -67,3 +73,180 @@ def deparallelize(self):
...
+            # loss = nn.functional.ctc_loss(
+            #     log_probs,
+            #     flattened_targets,
+            #     input_lengths,
+            #     target_lengths,
+            #     blank=self.config.pad_token_id,
+            #     reduction=self.config.ctc_loss_reduction,
+            #     zero_infinity=self.config.ctc_zero_infinity,
+            # )
+            loss = loss_fn(log_probs.float(), labels, input_lengths, target_lengths)
```
Comment (katalinic-gc) [inline optimum/graphcore/models/hubert/modeling_hubert.py:245]: potentially redundant cast to float as log_probs should already be float

File: optimum/graphcore/models/hubert/modeling_hubert.py [hunk 5]
```diff
@@ -67,3 +73,180 @@ def deparallelize(self):
...
+                self._add_begin_block(layer, name, ipu_id=layer_ipu[i])
+
+            logger.info("---------------------------------------")
+
+    def deparallelize(self):
+        """
+        Undo the changes to the model done by `parallelize`.
+        """
+        super().deparallelize()
+        self.change_hubert_encoder_class(True)
```
Comment (katalinic-gc) [inline optimum/graphcore/models/hubert/modeling_hubert.py:144]: in parallelize the feature encoder gets frozen; does any "unfreezing" need to happen here?

File: optimum/graphcore/models/hubert/modeling_hubert.py [hunk 6]
```diff
@@ -67,3 +73,180 @@ def deparallelize(self):
...
+            restore: whether to restore the encoder to its original version or not.
+        """
+        if self.config.do_stable_layer_norm:
+            new_cls = HubertEncoderStableLayerNorm if restore else IPUHubertEncoderStableLayerNorm
+        else:
+            new_cls = HubertEncoder if restore else IPUHubertEncoder
+        self.hubert.encoder.__class__ = new_cls
+
+    def _add_begin_block(self, module, name, ipu_id):
+        poptorch.BeginBlock(module, name, ipu_id)
```
Comment (katalinic-gc) [inline optimum/graphcore/models/hubert/modeling_hubert.py:93]: if I'm not mistaken in most cases we assign the return value of the BeginBlock, is this approach ok? cleaner this way if so

Comment (jimypbr) [reply optimum/graphcore/models/hubert/modeling_hubert.py:92]: yes this way works as well

**Pull request merged**